### ISSUE TITLE
Illegal argument in isolate message : (object is a regular Dart Instance)" - Dart 1.22-dev.5 breaks pkg/test 

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#28287]
See https://travis-ci.org/dart-lang/shelf/jobs/188931055#L690

"Failed to load "test/message_change_test.dart": Invalid argument(s): Illegal argument in isolate message : (object is a regular Dart Instance)"

Digging in a bit, but this seems gnarly/bad/etc

Reproduction:

- Sync down https://github.com/dart-lang/shelf
- cd into `shelf`
- `pub get`
- `pub run test`

**Comment by lrhn** [comment]
If it appeared when I made platform library methods generic, maybe (just guessing) there could be a problem of a generic method tear-off no longer being considered compile-time constant and sendable over send-ports. Or maybe it's completely unrelated.

**Comment by kevmoo** [comment]
I just tried to JSON-encode the message – and it worked.

So I'm thinking it just something weird...please hold...

**Comment by kevmoo** [comment]
After digging and bisecting, @lrhn it did turn out that 44594277e16f2b523eae8c8bc9ac60fd09bb930a was the culprit

**Comment by lrhn** [comment]
Since the syntax change from comment-based-generics to real generic syntax shouldn't affect semantics of the VM at all, this is a VM bug.
The error message comes from some object not being sendable through a `SendPort`, and since it used to work, it must be some new object that has changed to no longer be sendable. 
I'm guessing it's a closure of some sort, that didn't use to be a closure - you can send top-level/static functions to similar isolates, but not closures that capture part of their environment, so maybe a top-level generic function now thinks it's a closure because it contains the type variable. But that's just a guess based on what was actually changed, I have no actual idea what the problem is :)

**Comment by lrhn** [comment]
I've tried adding some debug prints just before the error is thrown in the VM (`raw_object_snapshot.cc` line 122) and the error appears because the test:
```dart
if (writer->can_send_any_object() ||
    writer->AllowObjectsInDartLibrary(ptr()->library_)) {
```
fails both tests with the library URI being `package:test/src/backend/group_entry.dart`. That's slightly surprising because the only thing in the library is an *abstract* class, so whatever object is being sent, it's not originally from that library

**Comment by a-siva** [comment]
why doesn't 'pub --trace run test' produce any trace of what dart command is being run. Pretty unhelpful diagnostics from pub, how does one get at the actual dart command being run by pub?

**Comment by a-siva** [comment]
[0x000000000067e3c5] dart::DynamicAssertionHelper::Fail(char const*, ...)
  [0x00000000009ea52a] dart::RawClass::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009eaf4f] dart::RawType::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009ebd51] dart::RawTypeArguments::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009f3478] dart::RawGrowableObjectArray::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009f3920] dart::RawLinkedHashMap::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009f3920] dart::RawLinkedHashMap::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x0000000000973a24] dart::SnapshotWriter::ArrayWriteTo(long, long, long, dart::RawSmi*, dart::RawTypeArguments*, dart::RawObject**, bool)
  [0x00000000009f3151] dart::RawArray::WriteTo(dart::SnapshotWriter*, long, dart::Snapshot::Kind, bool)
  [0x00000000009717a9] dart::SnapshotWriter::WriteObject(dart::RawObject*)
  [0x0000000000974569] dart::MessageWriter::WriteMessage(dart::Object const&)
  [0x0000000000a20823] dart::BootstrapNatives::DN_SendPortImpl_sendInternal_(_Dart_NativeArguments*)

**Comment by kevmoo** [comment]
`dart --packages=.packages ~/.pub-cache/hosted/pub.dartlang.org/test-0.12.15+12/bin/test.dart`

You can find the path to pkg/test in the `.packages` file – just replace `/lib/` with `/bin/test.dart`

**Comment by keertip** [comment]
This is now blocking the internal dev5.0 roll.

**Comment by nex3** [comment]
`test` only sends JSON-safe objects over the isolate API—we use the same code path for browser tests which use JSON, so if anything else were being sent it would break them. This suggests that whatever the problem object is, it looks fine to the JSON converter. I'd guess it's a `Map` or `List` of some kind.

**Comment by nex3** [comment]
Actually, judging by the CL Kevin bisected, I have a more specific guess: the Isolate API is rejecting collection types with reified generics that refer to unserializable classes. It probably always has, and we just never passed in a reified type before since it was so difficult to retain reification in spec mode. 

This seems like a bug in the Isolate API, not only because it's a breaking change but because it breaks the invariant that anything JSON-safe is Isolate-safe as well.

**Comment by a-siva** [comment]
As Lasse points out the error occurs because we seem to be sending a type whose class is 'GroupEntry' from the library 'package:test/src/backend/group_entry.dart' across to an isolate that has been spawned using spawnUri, this is not allowed by the VM and hence the exception.
What has changed recently in pkg/test or pub that makes it send this type object across isolates?

**Comment by kevmoo** [comment]
> What has changed recently in pkg/test or pub that makes it send this type object across isolates?

Nothing. I've tried this w/ a number of older version of pkg/test with no issue.

The key to kicking this off is change 44594277e16f2b523eae8c8bc9ac60fd09bb930a in the SDK.

I could imagine pkg/test could be fixed here – but from an outside observer, this is a regression in the SDK behavior, not pkg/test.

**Comment by a-siva** [comment]
The API document says "The content of message can be: primitive values (null, num, bool, double, String), instances of SendPort, and lists and maps whose elements are any of these. List and maps are also allowed to be cyclic."

**Comment by a-siva** [comment]
This test has ended up with an object in a Map that has a type object of library:'package:test/src/backend/group_entry.dart' and class:GroupEntry, this is not considered a primitive value and hence the exception.

**Comment by kevmoo** [comment]
https://github.com/dart-lang/test/blob/62cf174635490fe58b4d7e371f1bb3accbf333f7/lib/src/backend/group.dart#L51

If you remove the generic method from this line – for whatever reason – it works again.

```diff
@@ -48,7 +48,7 @@ class Group implements GroupEntry {

   Group(this.name, Iterable<GroupEntry> entries, {Metadata metadata,
           this.trace, Test this.setUpAll, Test this.tearDownAll})
-      : entries = new List<GroupEntry>.unmodifiable(entries),
+      : entries = new List.unmodifiable(entries),
         metadata = metadata == null ? new Metadata() : metadata;

   Group forPlatform(TestPlatform platform, {OperatingSystem os}) {
```

**Comment by kevmoo** [comment]
A better fix
`group.entries.map(...)` returns a `MappedIterable<GroupEntry, dynamic>`

Calling `.toList()` just returns what seems to be a normal Dart list – wondering if the `GroupEntry` type is getting leaked into it somehow...

```diff
diff --git a/lib/src/runner/remote_listener.dart b/lib/src/runner/remote_listener.dart
index d245a97..f496ced 100644
--- a/lib/src/runner/remote_listener.dart
+++ b/lib/src/runner/remote_listener.dart
@@ -134,11 +134,11 @@ class RemoteListener {
       "trace": group.trace?.toString(),
       "setUpAll": _serializeTest(channel, group.setUpAll, parents),
       "tearDownAll": _serializeTest(channel, group.tearDownAll, parents),
-      "entries": group.entries.map((entry) {
+      "entries": new List.from(group.entries.map((entry) {
         return entry is Group
             ? _serializeGroup(channel, entry, parents)
             : _serializeTest(channel, entry, parents);
-      }).toList()
+      }))
     };
   }
```

**Comment by lrhn** [comment]
The `toList` call preserves the type parameter of the receiving `Iterable`.
There are two problems here:
- I agree with Natalie that this is something we should fix - We should allow *any* `List` to be sent as "JSON-like" if its elements are sendable, and always drop the type parameter. When we send objects to "similar" isolates (created using `Isolate.spawn`) then it's fine to preserve the type parameter, but for "foreign" isolates (created using `Isolate.spawnUri`) we should only care about the `List`/`Map` structure, not the type parameters.
- Something changed due to switching to generic syntax from comment-based generics. Nothing should change, so something is wrong. It could be that I have introduced the wrong type parameter somewhere (maybe because the comments were wrong), or that the VM doesn't ignore the new syntax completely. I'll try to see if I can find a bug in that.

**Comment by dgrove** [comment]
Any further updates?

**Comment by a-siva** [comment]
The VM never supported sending of any List that is sending objects that implemented Lists or any Map that is objects that implemented Maps (what is being referred to as "JSON-like"). It was restricted to Lists and Maps from the core libraries and when sending these objects we have always send the type arguments also.

Clearly something has changed here that has caused a Map to become a map of type <String, List<GroupEntry>> and fixing that would be the most expedient thing to do here.

**Comment by nex3** [comment]
> The VM never supported sending of any List that is sending objects that implemented Lists or any Map that is objects that implemented Maps (what is being referred to as "JSON-like"). It was restricted to Lists and Maps from the core libraries and when sending these objects we have always sent the type arguments also.

Orthogonal to the issue reported here, this is a problem. It's opaque to users whether their types come from the core library or not, and they have a reasonable expectation that anything safe to JSON-encode is safe to send over an isolate. In addition, the documentation says it allows "lists and maps"—user types that implement these are still lists and maps.

I can open a separate issue for this if you'd like, but it needs to be addressed.

**Comment by a-siva** [comment]
I think the type leak is happening here in sdk/lib/collection/list.dart
`Iterable<T> map<T>(T f(E element)) => new MappedListIterable<E, T>(this, f);`
The old code used to be 
`Iterable/*<T>*/ map/*<T>*/(/*=T*/ f(E element)) => new MappedListIterable/*<E, T>*/(this, f);`
The old code used to created untyped MappedListIterable objects the new one will create
a `<E, dynamic> MappedListIterable` objects where E comes from the class.

I changed this code to  
`Iterable<T> map<T>(T f(E element)) => new MappedListIterable<dynamic, T>(this, f);`
and the test works.

**Comment by floitschG** [comment]
My expectation was also, that Lists and Maps would be serialized (as if they were encoded as JSON).

However, dart2js actually behaves the same. (And I wrote the serializer there...)

In the short term, we should simply fix the code that calls the `send` to become a "native" List/Map again. Longer term we should change our cross-URI communication code to simulate a JSON encoding that supports SendPorts.

**Comment by nex3** [comment]
> I changed this code to  
> `Iterable<T> map<T>(T f(E element)) => new MappedListIterable<dynamic, T>(this, f);`
> and the test works.

How confident are we that changing the reified type from `E` to `dynamic` won't break strong mode at runtime?

Also, there's still a mystery here of why `MappedListIterable<E, T>.toList()` returns a `List` that holds onto a reference of `E`. It should eagerly create a list that doesn't retain any references to the iterable that created it.

> In the short term, we should simply fix the code that calls the `send` to become a "native" List/Map without generic types, again.

@floitschG None of our code is currently sending non-native collections; the bug comes from the issue @a-siva pointed out above. I just want to remove sharp edges from the API.

**Comment by dgrove** [comment]
Thanks - can you all please ensure that a fix for this goes in today?

**Comment by a-siva** [comment]
@nex3 regarding your comment about supporting any list/map yes please open an enhancement request for that against the isolate API specification. I think this API needs clearer specification as people seem to have different assumptions than what is actually implemented.

**Comment by floitschG** [comment]
@a-siva: completely agree for the new bug. At least we seem to have consistent behavior (VM and dart2js behaving the same).

@nex3: you are right. The `toList()` should definitely lose the "from" type.

I tried simple examples, but couldn't reproduce so far. Just creating an unmodifiable list and calling `toList` on it, doesn't trigger the bug.

**Comment by a-siva** [comment]
Regarding a fix, maybe @floitschG or somebody from the library team should look at the changed line that was introduced in https://github.com/dart-lang/sdk/commit/44594277e16f2b523eae8c8bc9ac60fd09bb930a and determine if that needs to change or the user code that does the send needs to change.

**Comment by floitschG** [comment]
@dgrove: since we haven't figured out what the problem is, we can't really commit to a fix yet.

@a-siva: The line looks good.

I will start downloading the code from the first comment, but does anyone have a smaller repro?

**Comment by floitschG** [comment]
I have continued my investigation, but I'm really leaning towards a VM bug now.

I changed the test-package to print the runtime-type, and it states that the list has a dynamic type argument. I even use `dart:mirrors` to make sure that there wasn't any trickery going on.

This means, that the list returned by `group.entries.map(...).toList()` claims to be a `List<dynamic>` (by printing its `.runtimeType` and by using `dart:mirrors`), but eventually leads to troubles when being serialized.

It's *not* the entries in the list, since a simple `new List(...)` or a `.map((x) => x)` will make the error go away.

I don't know how @kevmoo figured out that the `new List.from` would make a difference, but it saved me a lot of time. Thanks!

Here is my diff:
``` patch
diff --git a/lib/src/runner/remote_listener.dart b/lib/src/runner/remote_listener.dart
index d245a97..5d75a83 100644
--- a/lib/src/runner/remote_listener.dart
+++ b/lib/src/runner/remote_listener.dart
@@ -17,6 +17,7 @@ import '../backend/test.dart';
 import '../backend/test_platform.dart';
 import '../util/remote_exception.dart';
 import '../utils.dart';
+import 'dart:mirrors';
 
 class RemoteListener {
   /// The test suite to run.
@@ -127,6 +128,17 @@ class RemoteListener {
   Map _serializeGroup(MultiChannel channel, Group group,
       Iterable<Group> parents) {
     parents = parents.toList()..add(group);
+    var t = group.entries.map((entry) {
+        return entry is Group
+            ? _serializeGroup(channel, entry, parents)
+            : _serializeTest(channel, entry, parents);
+      }).toList();
+    var m = reflect(t);
+    assert(m.type.toString() == "ClassMirror on '_GrowableList'");
+    assert(m.type.typeArguments.toString() == "[TypeMirror on 'dynamic']");
+    // Uncomment one of the following lines to make the error go away.
+    // t = new List.from(t);
+    // t = t.map((x) => x).toList();
     return {
       "type": "group",
       "name": group.name,
@@ -134,11 +146,7 @@ class RemoteListener {
       "trace": group.trace?.toString(),
       "setUpAll": _serializeTest(channel, group.setUpAll, parents),
       "tearDownAll": _serializeTest(channel, group.tearDownAll, parents),
-      "entries": group.entries.map((entry) {
-        return entry is Group
-            ? _serializeGroup(channel, entry, parents)
-            : _serializeTest(channel, entry, parents);
-      }).toList()
+      "entries": t
     };
   }
```

**Comment by a-siva** [comment]
After some debugging on our side inside gdb it does look like a VM issue. I have to discuss a bit more with Regis on a fix for this. A potential fix is only possible tomorrow.

On a side note, test/pkg seems to be part of the sdk why do we not run these tests as part of the regression tests on every build?

**Comment by nex3** [comment]
> On a side note, test/pkg seems to be part of the sdk why do we not run these tests as part of the regression tests on every build?

It's not really a part of the SDK; it (like many other Dart packages) is pulled in from GitHub using the DEPS file. Those packages notionally have their own CI, and aren't tested as part of the SDK's suite.

**Comment by lrhn** [comment]
On Tue, Jan 10, 2017 at 10:18 PM, Siva <notifications@github.com> wrote:

> I think the type leak is happening here in sdk/lib/collection/list.dart
> Iterable<T> map<T>(T f(E element)) => new MappedListIterable<E, T>(this,
> f);
> The old code used to be
> Iterable/*<T>*/ map/*<T>*/(/*=T*/ f(E element)) => new
> MappedListIterable/*<E, T>*/(this, f);
>
Good find.
That change is actually fixing a bug in the original code, the original
should have been:

Iterable/*<T>*/ map/*<T>*/(/*=T*/ f(E element)) => new
MappedListIterable<E, dynamic/*T*/>(this, f);

It shouldn't leak the type `E` into anything, though. The result of
`MappedListIterable<A, B>.toList()` should be a List<B>, not a List<A>.
The exact code is:

/**
 * Specialized alternative to [MappedIterable] for mapped [List]s.
 *
 * Expects efficient `length` and `elementAt` on the source iterable.
 */
class MappedListIterable<S, T> extends ListIterable<T> {
  final Iterable<S> _source;
  final _Transformation<S, T> _f;

  MappedListIterable(this._source, T this._f(S value));

  int get length => _source.length;
  T elementAt(int index) => _f(_source.elementAt(index));
}

All other methods are inherited from `ListIterable<T>` which doesn't even
know about `S` (it's defined entirely in terms of `length` and `elementAt`,
it can't see `_source` or `S` at all).


> The old code used to created untyped MappedListIterable objects the new
> one will create
> a <E, dynamic> MappedListIterable objects where E comes from the class.
>
True, but barring a bug in the VM, that shouldn't be able to leak E.


> I changed this code to
> Iterable<T> map<T>(T f(E element)) => new MappedListIterable<dynamic,
> T>(this, f);
> and the test works.
>
Ick. That does sound like a bug in the VM's treatment of type parameters
(not that I can reproduce it, but I think my argument against the first
type parameter leaking is sound (feel free to prove me wrong, so I can fix
it :).

/L
-- 
Lasse R.H. Nielsen - lrn@google.com
'Faith without judgement merely degrades the spirit divine'
Google Denmark ApS - Frederiksborggade 20B, 1 sal - 1360 København K
- Denmark - CVR nr. 28 86 69 84

**Comment by a-siva** [comment]
> It's not really a part of the SDK; it (like many other Dart packages) is pulled in from GitHub using the DEPS file. Those packages notionally have their own CI, and aren't tested as part of the SDK's suite.

What about the tree I see in our build tree labelled 'pub-pkg', would it make sense to add it there?
The reason I bring it up is this subtle regression was introduced about 8 days ago and we did not discover it till couple of days back.

**Comment by whesse** [comment]
The error does show up in the package testers in that group on the main waterfall.  That exact error shows up in https://uberchromegw.corp.google.com/i/client.dart/builders/pkg-linux-be/builds/5993/steps/third_party%20pkg%20tests/logs/stdio

Packages in third_party that we want to test with every SDK commit are put in third_party/pkg_tested instead of third_party/pkg.  Test is not currently there, but other packages that use test are there, and the error showed up.  All packages in pkg (not third_party/pkg) also have all their unit tests run there.

I have in the past made these builders green, and been filing issues and fixes against packages like test and dart_style to fix failing tests there, but it seems that the package owners don't work on keeping them green, because they prefer their own testing methods.

**Comment by matanlurey** [comment]
I think if we are relying on something for the SDK or a Dart release (like it or not, pkg/test and pkg/dart_style *are* relied on) they should stay green on the bots so the SDK-team has visibility into breakages. My 2 cents.

**Comment by keertip** [comment]
The issue with the pub-pkg bots is that it does not have a designated team/owner, so failures sometimes don't get triaged. Also, they are red quite a bit, and so are assumed already failing on commits and not looked at.

**Comment by dgrove** [comment]
Do we have a plan for getting this fixed? The roll is still blocked.

**Comment by a-siva** [comment]
fix is being code reviewed and should land today

**Comment by dgrove** [comment]
Thanks! Can you please ask Bill to do a push as soon as this in landed?

**Comment by kevmoo** [comment]
Fixed in f634059f6f321dc6abbb41997405695320eae0a6

Verified in 1.22.0-dev.7.0

Thanks!
